### PR TITLE
Add workflow function to user

### PR DESCRIPTION
**Pull request by itkg-nanne**

### CODE REVIEW

**Comment by itkg-ngilain** [comment]
:+1:

**Comment by nicolasThal** [comment]
There are some test missing

File: UserAdminBundle/Form/DataTransformer/ContentTypeToAuthorizationTransformer.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+use Doctrine\Common\Collections\ArrayCollection;
+use OpenOrchestra\ModelBundle\Repository\ContentTypeRepository;
+use OpenOrchestra\UserBundle\Model\AuthorizationInterface;
+
+/**
+ * Class ContentTypeToAuthorizationTransformer
+ */
+class ContentTypeToAuthorizationTransformer implements DataTransformerInterface
+{
+
```
Comment (nicolasThal) [inline UserAdminBundle/Form/DataTransformer/ContentTypeToAuthorizationTransformer.php]: empty line tu suppress

**Comment by itkg-nanne** [comment]
mieux comme ça, @nicolasThal ?

File: UserAdminBundle/Resources/config/form.yml [hunk 2]
```diff
@@ -1,16 +1,15 @@
 parameters:
-    open_orchestra_user.type.user.class: OpenOrchestra\UserAdminBundle\Form\Type\UserType
-    open_orchestra_user.type.registration_user.class: OpenOrchestra\UserAdminBundle\Form\Type\RegistrationUserType
-
+    open_orchestra_user_admin.type.user.class: OpenOrchestra\UserAdminBundle\Form\Type\UserType
+    open_orchestra_user_admin.type.registration_user.class: OpenOrchestra\UserAdminBundle\Form\Type\RegistrationUserType
 services:
-    open_orchestra_user.type.user:
-        class: %open_orchestra_user.type.user.class%
+    open_orchestra_user_admin.type.user:
```
Comment (nicolasThal) [inline UserAdminBundle/Resources/config/form.yml]: do not forget to put this change in the changelog

File: WorkflowFunctionAdminBundle/Controller/Admin/WorkflowRightController.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     *
+     * @Config\Security("has_role('ROLE_ACCESS_USER')")
+     *
+     * @return Response
+     */
+    public function formAction(Request $request, $userId)
+    {
+        $user = $this->get('open_orchestra_user.repository.user')->find($userId);
+        $workflowRight = null;
+        if (null !== $user) {
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowRightController.php]: I think that if the user is Null you should throw an exception

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace OpenOrchestra\WorkflowFunctionAdminBundle\Manager;
+
+use Doctrine\ODM\MongoDB\DocumentManager;
+use OpenOrchestra\ModelBundle\Repository\ContentTypeRepository;
+use OpenOrchestra\WorkflowFunction\Model\WorkflowRightInterface;
+
+
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: too many empty lines

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        $this->authorizationClass = $authorizationClass;
+        $this->workflowRightClass = $workflowRightClass;
+    }
+
+    /**
+     * @param WorkflowRightInterface|null $workflowRight
+     *
+     * @return WorkflowRightInterface
+     */
+    public function clean($workflowRight)
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: you should authorize null parameter in the signature

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+use Doctrine\ODM\MongoDB\DocumentManager;
+use OpenOrchestra\ModelBundle\Repository\ContentTypeRepository;
+use OpenOrchestra\WorkflowFunction\Model\WorkflowRightInterface;
+
+
+
+/**
+ * Class WorkflowRightManager
+ */
+class WorkflowRightManager
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: i cannot find the test for this class

File: WorkflowFunctionAdminBundle/Controller/Admin/WorkflowRightController.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+     * @Config\Route("/form/{userId}", name="open_orchestra_backoffice_workflow_right_form")
+     * @Config\Method({"GET", "POST"})
+     *
+     * @Config\Security("has_role('ROLE_ACCESS_USER')")
+     *
+     * @return Response
+     */
+    public function formAction(Request $request, $userId)
+    {
+        $user = $this->get('open_orchestra_user.repository.user')->find($userId);
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowRightController.php]: the only parameter you need is the userId, why would you get the whole user?

Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowRightController.php]: https://github.com/open-orchestra/open-orchestra-workflow-function-bundle/pull/3/files#diff-dee2253013ab630072f8e9357539772fR22

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php (hunk 5 above)
Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: Isn't what i did?

Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: clean($workflowRight = null) ?

**Comment by itkg-nanne** [comment]
Is it ok for you before i add tests? @nicolasThal

File: WorkflowFunctionAdminBundle/Form/Type/AuthorizationType.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        $builder->add('referenceId', 'hidden');
+        $builder->add('workflowFunctions', 'orchestra_workflow_function', array(
+            'label' => false,
+            'required' => false
+        ));
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        //$view->vars['label'] = $view->vars['value']->getName();
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Form/Type/AuthorizationType.php]: if this method is empty, you can supress it

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    protected $workflowRightRepository;
+
+    protected $authorizationWorkflowRightManager;
+
+    protected $workflowRightClass;
+
+    /**
+     * Constructor
+     *
+     * @param ContentTypeRepositoryInterface    $contentTypeRepository
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: a parameter is missing

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * @param string $userId
+     *
+     * @return WorkflowRightInterface
+     */
+    public function loadOrGenerateByUser($userId)
+    {
+        $contentTypes = $this->contentTypeRepository->findAllByDeletedInLastVersion();
+        $reference = new Reference();
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: can you use a parameter for this class?

File: WorkflowFunctionAdminBundle/Resources/config/resolve_document.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+resolve_target_documents:
+        OpenOrchestra\WorkflowFunction\Model\ReferenceInterface: OpenOrchestra\ModelBundle\Document\ContentType
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Resources/config/resolve_document.yml]: Why this relation ? 

Isn't there a Reference class that implements the ReferenceInterface

**Comment by nicolasThal** [comment]
you can add some tests :)

**Comment by itkg-nanne** [comment]
done @nicolasThal

File: UserAdminBundle/Form/Type/UserType.php [hunk 12]
```diff
@@ -15,7 +15,7 @@ class UserType extends AbstractType
     protected $class;
 
     /**
-     * @param string              $class
+     * @param string  $class
```
Comment (nicolasThal) [inline UserAdminBundle/Form/Type/UserType.php]: there are two spaces instead of one

File: UserAdminBundle/Resources/config/form.yml (hunk 2 above)
Comment (nicolasThal) [inline UserAdminBundle/Resources/config/form.yml]: did you put it in the changelog?

File: WorkflowFunctionAdminBundle/Form/Type/AuthorizationType.php [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add('referenceId', 'hidden');
+        $builder->add('workflowFunctions', 'orchestra_workflow_function', array(
+            'label' => false,
+            'required' => false
+        ));
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options)
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Form/Type/AuthorizationType.php]: method comment

File: WorkflowFunctionAdminBundle/Tests/Form/AuthorizationTypeTest.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+namespace OpenOrchestra\BackofficeBundle\Tests\Form\Type;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use OpenOrchestra\WorkflowFunctionAdminBundle\Form\Type\AuthorizationType;
+use Phake;
+
+/**
+ * Description of AuthorizationTypeTest
+ */
+class AuthorizationTypeTest extends \PHPUnit_Framework_TestCase
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Tests/Form/AuthorizationTypeTest.php]: did you forgot to commit the test for the buildView method?

File: WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+use OpenOrchestra\WorkflowFunctionBundle\Document\Reference;
+use OpenOrchestra\ModelInterface\Repository\ContentTypeRepositoryInterface;
+use OpenOrchestra\WorkflowFunction\Repository\WorkflowRightRepositoryInterface;
+
+/**
+ * Class WorkflowRightManager
+ */
+class WorkflowRightManager
+{
+
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Manager/WorkflowRightManager.php]: there is no need to let so many empty lines

**Comment by nicolasThal** [comment]
I also think that at least one test is broken

File: WorkflowFunctionAdminBundle/Tests/Form/AuthorizationTypeTest.php (hunk 14 above)
Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Tests/Form/AuthorizationTypeTest.php]: can't do it because of $view['']

**Comment by itkg-alavieille** [comment]
:+1

**Pull request merged**